### PR TITLE
Add `unique_id` attribute to Layer to uniquely identify layers

### LINKED ISSUE #5229: Layers should have an attribute to uniquely identify them
## 🚀 Feature

A way to uniquely identify layers.

## Motivation

@cudmore and I are working on an undo-redo plugin for napari layers. As part of this we are storing undo and redo stacks for each napari layer. However, we need to have a way to uniquely identify layers.

## Pitch

If each layer has an `id` attribute that is unique (by using uuid.uuid4) and generated once when the layer is created, then we can easily use this attribute to identify each layer and their data for undo and redo operations.

## Alternatives

We could use the `name` attribute, but that can change and we'll need to track that. We have also tried using `id(layer)` which gives us the layer's address, but this may not stay unique during the lifetime of the program.

## Additional context

Created https://github.com/napari/napari/pull/5230 to add the id attribute.

### PR DESCRIPTION
**Pull request by VasudhaJha**
Closes #5229

# Description
Add `layer.unique_id` guaranteed to be unique and hashable (capable of equality comparison) across the lifetime of a viewer. Currently implemented using `UUID` but this **may** change.

## Type of change
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by liu-ziyang** [comment]
Reading the original issue discussion, I hesitate to add UUID to layers. 
Though I don't think hash would work in this case (I don't yet find evidence that hash would be consistent upon a layer data change, and even if it does today, it is not a guaranteed behavior across versions on what parameters are included in the hash)

I also hesitate to use the id function directly as according to the doc the implementation is based on memory address and could potentially have conflicts between objects with non-overlapping lifecycle, I can imaging corner case bugs where the id collision happens if we don't carefully clean up the mapping based on object lifecycles.

Before we add the UUID tracking, which does seem to address the problems above, I wonder if the suggested weak reference would be a better approach. The added UUID seems to be one more thing that needs to be maintained and tracked. Plus any reference would involve either looping through all layers to find it or building yet another data structure tracking all UUIDs. Is there any reason we can't use weak references of layers?

**Comment by jni** [comment]
@VasudhaJha sorry that this fell through the cracks. Actually I think UUID is a fine solution here. Looking at the original issue and @potating-potato's comment, I agree that hash is not the right tool for this. But UUID is (imho). @Czaki what do you think about this?

**Comment by jni** [comment]
I vaguely think using `.uuid` as the name is better, so it doesn't conflict with the built-in `id()`. (I know it's an attribute so not strictly a conflict, but e.g. syntax checkers still highlight it.)

**Comment by jni** [comment]
Discussed this at the PR party. Came to the following decisions:
- @DragaDoncila will pick this up and push to this branch to update the PR
- The id attribute will *for now* be implemented as UUID, but the public API will only guarantee that it is unique for the lifetime of the program — the implementation could change to e.g. unique integers in the future. Users should *not* rely on this being a UUID.

File: napari/layers/base/base.py [hunk 1]
```diff
@@ -252,7 +253,7 @@ def __init__(
         # Needs to be imported here to avoid circular import in _source
         from .._source import current_source
 
-        self.id = uuid.uuid4()
+        self.unique_id = self._get_unique_id()
```
Comment (Czaki) [inline napari/layers/base/base.py]: Maybe we should make this attribute private and add read_only property for it?

Comment (DragaDoncila) [reply napari/layers/base/base.py]: @Czaki totally agree it should be read only, will update this. However, I'm not sure why we want to make it private? Based on the initial request in #5229 it seems like plugins may want to use it, and having it as a private attribute would discourage that I assume. Or are you suggesting this as a temporary measure, and we'd eventually make it public?

File: napari/layers/_tests/test_layer_attributes.py [hunk 2]
```diff
@@ -134,6 +136,19 @@ def test_get_value_3d_view_of_2d_image(ImageClass):
     _check_subpixel_values(layer, val_dict)
 
 
+@pytest.mark.parametrize('Layer, data, _', layer_test_data)
+def test_layer_unique_id(Layer, data, _):
+    layer = Layer(data)
+    assert layer.unique_id is not None
+    assert type(layer.unique_id) is uuid.UUID
```
Comment (jni) [inline napari/layers/_tests/test_layer_attributes.py]: Tests should ideally only test that the public API promises are maintained. In this case, this is testing an implementation detail — this test should only test that the type is hashable.

File: napari/layers/base/base.py (hunk 1 above)
Comment (jni) [reply napari/layers/base/base.py]: Yes I think what we discussed is that this should be a property annotated with the type `collections.abc.hashable`

**Review by jni — CHANGES_REQUESTED** [review]
Hey @DragaDoncila! This is close but I'm requesting two changes, based on last night's discussion:

- test should only test that id is hashable, not UUID
- change layer.unique_id to read-only property with proper type annotation

File: napari/layers/base/base.py [hunk 3]
```diff
@@ -499,6 +504,10 @@ def mode(self, mode):
 
         self.events.mode(mode=str(mode))
 
+    def _get_unique_id(self):
+        """Unique id of the layer."""
+        return uuid.uuid4()
```
Comment (jni) [inline napari/layers/base/base.py]: Extra comment: there should be a guard here to make sure that this is only run once. I guess rolling this into the read-only property would be good!

**Review by jni — APPROVED** [review]

**Comment by psobolewskiPhD** [comment]
@DragaDoncila @jni What's the status of this? still seems useful...

**Comment by jni** [comment]
Thanks @VasudhaJha and sorry that it took us so long to get this over the line!!! 🙏

**Pull request merged**